### PR TITLE
Only quit from menus on Mac when Cmd+Q is pressed

### PR DESCRIPTION
**Pull request by ghost**
If there's a server running and it's not finished the game yet, then we ignore the request to exit the application when Cmd + Q is pressed.  This is particularly important on Mac because Cmd + Q is quit the game's entire process and Cmd + 1 is assign the units to the first group.  Two very close together keyboard combinations with that latter being very heavily used in the game.

The conditions I've set for this command to be ignored appear to apply on Single player missions and in the map editor.  I wasn't able to test the multiplayer, but I'm confident it would work there too.

### CODE REVIEW

**Review by abcdefg30 — COMMENTED** [review]
I cannot close the game via the close button (X at the top window bar) any more when ingame, which imo is a regression.

**Comment by abcdefg30** [comment]
On the other hand I suppose that is part of the whole point of this PR.

**Comment by deleted-user-1** [comment]
> I cannot close the game via the close button (X at the top window bar) any more when ingame, which imo is a regression.

> On the other hand I suppose that is part of the whole point of this PR.

This is reverse spacebar-heating, IMO.

**Comment by pchote** [comment]
It is a fair concern, IMO - the PR is in fact not just suppressing Cmd + Q on mac but *all* OS-requested termination events, on *all* OSes. This is likely to have unexpected consequences.

The [SDL docs](https://wiki.libsdl.org/SDL_EventType#SDL_QUIT) references a `SDL_WINDOWEVENT_CLOSE` - it may be possible to switch to this to keep the window button handling while removing the hardcoded(?) sdl cmd+q behaviour.

File: OpenRA.Game/Game.cs [hunk 1]
```diff
@@ -855,6 +855,14 @@ static RunStatus Run()
 			return state;
 		}
 
+		public static void ExitIfNotRunningGame()
+		{
+			if (server == null || !server.IsStartedOrWaiting())
```
Comment (pchote) [inline OpenRA.Game/Game.cs]: This won't work in multiplayer games. You may be able to check something inside `OrderManager` instead.

**Comment by pchote** [comment]
IMO the optimum behaviour here would be suppress the hotkey during gameplay, but allow it when in menus (including the ingame menu) - matching the behaviour that we want for locking the cursor to the window.

**Review by deleted-user-1 — CHANGES_REQUESTED** [review]
This blocks SIGINT from terminal on Linux, which is the quickest way to exit to test changes across mods for example.

**Comment by deleted-user-1** [comment]
The "Stop"-Button in the vscode-debugging tooling added in #18280 still works, but to tbh I'd like to be able to interrupt from terminal too. Personally I question that avoiding an erroneously issued command can justify blocking correct OS response for everyone who intentionally issues this command. To me it looks that Apple's keybinding is just poorly designed (compared to less error-prone ALT+F4 on Windows / Ubuntu) and can't see why users of other OS's shall pay the price for this.

**Comment by pchote** [comment]
Alt + F4 will also terminate the game. The problem is that Cmd + Q is just half a fingers press away from the Cmd + 1 control group, which makes it far too easy to trigger.

**Comment by deleted-user-1** [comment]
I see that it is a legitimate motivation but at the current state the IMO undesired side effects are overwhelming. Could this be made Apple and/or production-build only?

**Comment by deleted-user-1** [comment]
> Alt + F4 will also terminate the game. The problem is that Cmd + Q is just half a fingers press away from the Cmd + 1 control group, which makes it far too easy to trigger.

You can reassign this for OSX per application. https://www.quora.com/How-do-I-disable-Command-q-in-OSX

**Comment by tovl** [comment]
IMO this is basically like #15527 again: It is an issue with the window manager and not with OpenRA. It should and can be fixed by changing the appropriate settings in the window manager. This is actually even far less severe than #15527 because you have to actually mistype before it becomes a problem. We should just add an entry to the FAQ for it, like we did here: https://github.com/OpenRA/OpenRA/wiki/FAQ/_compare/184418d827de85a621d57a2c79d62bbbb70be220...7965d7d1ddddc97bb6ec286899813949a7f46b16

**Comment by pchote** [comment]
>  Could this be made Apple and/or production-build only?

I already left review comments above suggesting a potential solution.

**Comment by abcdefg30** [comment]
Tbf, I 100% agree that this is MacOS weirdness and not OpenRA, but I don't think having a mac specific workaround would be bad.

**Comment by ghost** [comment]
I've made this Mac specific and removed an incomplete attempt to work out whether the game was in session.  This is in accordance with the advice given by @pchote on IRC.  I think there would be value in making this prevention only apply when the game was actively being played, but I think the convenience of preventing Cmd + Q from ruining your game is of larger value than the inconvenience that you cannot use Cmd + Q at all in the game any more.

**Review by abcdefg30 — DISMISSED** [review]
Code changes lgtm and I didn't spot regressions on Windows.

**Comment by ghost** [comment]
The last commit attempts to readd conditional exit on a Mac when pressing Cmd + Q.  The new boolean I'm using here appears to be correct no matter how you entered a game and whether you're hosting or not.  I'm happy to remove the last commit if it's determined to not be needed, but it allows the game to continue to respond to Cmd + Q in situations where you're not playing the game or editing a map.  The requirement that Cmd be down when SDL_QUIT is raised means that all other forms of exiting the game work all the time.  So the title bar button works again without having to catch the Window Close event.  And you can Ctrl + C from the terminal to quit the game at any time.  If the Cmd key is down though, those other mechanisms will behave like Cmd +Q.

File: OpenRA.Game/Game.cs [hunk 2]
```diff
@@ -855,6 +855,12 @@ static RunStatus Run()
 			return state;
 		}
 
+		public static void ExitIfNotMidGame()
+		{
+			if (OrderManager.Connection.ConnectionState != ConnectionState.Connected)
```
Comment (pchote) [inline OpenRA.Game/Game.cs]: The problem with this approach is that Cmd+q will quit if you're in e.g. the multiplayer server browser, but not if you're in a skirmish lobby. This distinction feels pretty arbitrary, and it wouldn't be fair to scope creep this PR to require more advanced / intuitive logic. So IMO it would be better to just universally block it for now.

**Comment by ghost** [comment]
OK so here we have my final attempt at something that turned out to be quite a bit more complex than I first imagined 😄

**Comment by deleted-user-1** [comment]
Can we have a similar workaround for Linux window managers then (for ALT)?

**Review by deleted-user-1 — DISMISSED** [review]
Side-effects on Linux are eliminated.

File: OpenRA.Platforms.Default/Sdl2Input.cs [hunk 3]
```diff
@@ -73,7 +73,10 @@ public void PumpInput(Sdl2PlatformWindow device, IInputHandler inputHandler, int
 				switch (e.type)
 				{
 					case SDL.SDL_EventType.SDL_QUIT:
-						Game.Exit();
+						// On macOS, we'd like to restrict Cmd + Q from suddenly exiting the game.
+						if (!(Platform.CurrentPlatform == PlatformType.OSX && mods.HasModifier(Modifiers.Meta)))
```
Comment (abcdefg30) [inline OpenRA.Platforms.Default/Sdl2Input.cs]: This can be simplified by de-morgan: `if (Platform.CurrentPlatform != PlatformType.OSX || !mods.HasModifier(Modifiers.Meta))`

**Comment by deleted-user-1** [comment]
Not sure if relevant, this is from the [SDL docs](https://wiki.libsdl.org/SDL_EventType#SDL_QUIT):
> Applications that completely ignore Command-Q will fail Mac App Store certification.

**Comment by ghost** [comment]
> Not sure if relevant, this is from the [SDL docs](https://wiki.libsdl.org/SDL_EventType#SDL_QUIT):
> 
> > Applications that completely ignore Command-Q will fail Mac App Store certification.

This is why I wanted to make the Cmd + Q combination work outside of active games, but it was agreed we should leave that complexity for another day.

**Comment by pchote** [comment]
Not really relevant, OpenRA already fails certification due to not having a legal license agreement for the C&C IP.

File: OpenRA.Platforms.Default/Sdl2Input.cs (hunk 3 above)
Comment (ghost) [reply OpenRA.Platforms.Default/Sdl2Input.cs]: I agree that’s a valid refactor here but I disagree it makes the logic simpler. The code should be as easy as possible for maintainers to understand. “Not macOS with the meta key down” feels easier to understand than the proposed “Not macOS or not the meta key down”. If the extra brackets are your main concern, the logic inside them could be extracted to a well named variable before the `if` statement.

Comment (abcdefg30) [reply OpenRA.Platforms.Default/Sdl2Input.cs]: Personally I find the latter easier to parse since I don't have to negate at the end, but don't really have a strong opinion here. (The compiler will likely optimize this away regardless.)

Comment (pchote) [reply OpenRA.Platforms.Default/Sdl2Input.cs]: I also find the latter easier to parse, FWIW. But I also don't think this is particularly important either way.

Comment (deleted-user-1) [reply OpenRA.Platforms.Default/Sdl2Input.cs]: > The code should be as easy as possible for maintainers to understand.
> Personally I find the latter easier to parse
> I also find the latter easier to parse

Can you please apply the fixup then?

**Comment by deleted-user-1** [comment]
Fixup commit should be squashed.

**Comment by ghost** [comment]
Just checking in to see if this is now approved by everyone?

**Review by abcdefg30 — APPROVED** [review]
Still works as advertised.

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/49bee230158c9b66e13df127f855cee349dc97b7
